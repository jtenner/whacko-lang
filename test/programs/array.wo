export fn _start(): void {
  let a = new Array<i32>(usize(100));
}