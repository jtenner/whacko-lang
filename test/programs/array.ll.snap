; ModuleID = 'whacko'
source_filename = "whacko"
target triple = "wasm32-wasi"

declare void @__whacko_gc_initialize()

declare ptr @__whacko_gc_alloc(i32, i32)

declare void @__whacko_gc_store_field_ptr(i32, ptr, ptr, ptr)

define void @_start() gc "shadow-stack" {
entry:
  call void @__whacko_gc_initialize()
  %"variable-a-alloca~0" = alloca ptr, align 8
  call void @llvm.gcroot(ptr %"variable-a-alloca~0", ptr null)
  %"result~635" = call ptr @__whacko_gc_alloc(i32 2, i32 ptrtoint (ptr getelementptr ({ ptr, ptr, i32, i32, ptr, i32 }, ptr null, i32 1) to i32))
  call void @"callable:[std/array.wo]Array.constructor;methodType:classType:[std/array.wo]Array<i32Type>#(usizeType)=>voidType"(ptr %"result~635", i32 100)
  store ptr %"result~635", ptr %"variable-a-alloca~0", align 8
  ret void
}

define void @"callable:[std/array.wo]Array.__gc_visit;methodType:classType:[std/array.wo]Array<i32Type>#()=>voidType"(ptr %0) gc "shadow-stack" {
entry:
  %"generated-alloca~0" = alloca ptr, align 8
  call void @llvm.gcroot(ptr %"generated-alloca~0", ptr null)
  store ptr %0, ptr %"generated-alloca~0", align 8
  ret void
}

define void @"callable:[std/array.wo]Array.constructor;methodType:classType:[std/array.wo]Array<i32Type>#(usizeType)=>voidType"(ptr %0, i32 %1) gc "shadow-stack" {
entry:
  %"generated-alloca~0" = alloca ptr, align 8
  call void @llvm.gcroot(ptr %"generated-alloca~0", ptr null)
  store ptr %0, ptr %"generated-alloca~0", align 8
  %"generated-alloca~1" = alloca i32, align 4
  store i32 %1, ptr %"generated-alloca~1", align 4
  %"variable-overhead-alloca~2" = alloca i32, align 4
  %"592~inst" = load ptr, ptr %"generated-alloca~0", align 8
  %"gep~636" = getelementptr { ptr, ptr, i32, i32, ptr, i32 }, ptr %"592~inst", i32 0, i32 4
  %"593~inst" = load ptr, ptr %"gep~636", align 8
  %"597~inst" = load i32, ptr %"variable-overhead-alloca~2", align 4
  %"598~inst" = load i32, ptr %"generated-alloca~1", align 4
  %"610~inst" = mul i32 %"598~inst", 32
  %"611~inst" = add i32 %"597~inst", %"610~inst"
  %"616~inst" = call ptr @__whacko_gc_alloc(i32 1, i32 %"611~inst")
  %"gep~637" = getelementptr { ptr, ptr, i32, i32, ptr, i32 }, ptr %"592~inst", i32 0, i32 4
  call void @__whacko_gc_store_field_ptr(i32 1, ptr %"592~inst", ptr %"616~inst", ptr %"gep~637")
  %"620~inst" = load ptr, ptr %"generated-alloca~0", align 8
  %"gep~638" = getelementptr { ptr, ptr, i32, i32, ptr, i32 }, ptr %"620~inst", i32 0, i32 5
  %"621~inst" = load i32, ptr %"gep~638", align 4
  %"gep~639" = getelementptr { ptr, ptr, i32, i32, ptr, i32 }, ptr %"620~inst", i32 0, i32 5
  store i32 0, ptr %"gep~639", align 4
  ret void
}

; Function Attrs: nounwind
declare void @llvm.gcroot(ptr, ptr) #0

declare void @__whacko_gc_visit(ptr)

define void @__visit_whacko_object(ptr %0) {
"entry~640":
  %"typeGEP~642" = getelementptr { ptr, ptr, i32, i32 }, ptr %0, i32 2
  %1 = load i32, ptr %"typeGEP~642", align 4
  switch i32 %1, label %"unreachable~641" [
    i32 1, label %"classType:[std/buffer.wo]ArrayBuffer<>"
    i32 2, label %"classType:[std/array.wo]Array<i32Type>"
    i32 3, label %"classType:[std/String.wo]String<>"
  ]

"unreachable~641":                                ; preds = %"entry~640"
  unreachable

"classType:[std/buffer.wo]ArrayBuffer<>":         ; preds = %"entry~640"
  ret void

"classType:[std/array.wo]Array<i32Type>":         ; preds = %"entry~640"
  %"gep~643" = getelementptr { ptr, ptr, i32, i32, ptr, i32 }, ptr %0, i32 0
  %"load~644" = load ptr, ptr %"gep~643", align 8
  call void @__whacko_gc_visit(ptr %"load~644")
  call void @"callable:[std/array.wo]Array.__gc_visit;methodType:classType:[std/array.wo]Array<i32Type>#()=>voidType"(ptr %0)
  ret void

"classType:[std/String.wo]String<>":              ; preds = %"entry~640"
  ret void
}

attributes #0 = { nounwind }
